File: src/ralph_wiggum/commands/loop.py
"""Loop command - start a Ralph Wiggum iterative loop."""

from typing import Annotated

from cyclopts import Parameter
from rich.console import Console

from ralph_wiggum.core.loop_runner import run_loop
from ralph_wiggum.models.config import LoopConfig

console = Console()


def loop_cmd(
    prompt: Annotated[str, Parameter(help="Task prompt for the loop")],
    min_iterations: Annotated[
        int,
        Parameter(
            name=["--min-iterations", "-n"],
            help="Minimum iterations before checking completion",
        ),
    ] = 1,
    max_iterations: Annotated[
        int,
        Parameter(name=["--max-iterations", "-m"], help="Max iterations (0=unlimited)"),
    ] = 0,
    completion_promise: Annotated[
        str,
        Parameter(
            name=["--completion-promise", "-p"],
            help="Promise phrase to signal completion (default: COMPLETE)",
        ),
    ] = "COMPLETE",
    agent_name: Annotated[
        str | None,
        Parameter(name=["--agent", "-a"], help="Agent name (default: ralph-wiggum)"),
    ] = None,
) -> None:
    """Start a Ralph Wiggum iterative loop.

    The loop runs kiro-cli repeatedly with fresh sessions until:
    - Minimum iterations completed AND completion promise detected
    - Max iterations is reached
    - User interrupts with Ctrl+C

    Parameters
    ----------
    prompt
        Task description for the agent to work on.
    min_iterations
        Minimum iterations before checking for completion promise.
    max_iterations
        Maximum iterations before stopping (0=unlimited).
    completion_promise
        Phrase that signals task completion when wrapped in <promise> tags.
    agent_name
        Name of the Kiro agent to use (discovered from .kiro/agents/).

    Examples
    --------
    ralph loop "Build a REST API" --min-iterations 3 --max-iterations 20

    ralph loop "Fix the auth bug" -n 2 -m 10 -p "FIXED"
    """
    config = LoopConfig(
        prompt=prompt,
        min_iterations=min_iterations,
        max_iterations=max_iterations,
        completion_promise=completion_promise,
        agent_name=agent_name,
    )

    run_loop(config, console)

File: src/commands/loop.ts
/**
 * @fileoverview Loop command for Ralph Wiggum CLI.
 * Starts an iterative loop that runs kiro-cli until completion.
 * @module commands/loop
 */
import { log } from "@clack/prompts";
import pc from "picocolors";

import { runLoop } from "../core/loop-runner";
import { LoopConfigSchema } from "../schemas/config";

/**
 * CLI options for the loop command (raw string values from commander).
 */
interface LoopOptions {
	/** Minimum iterations before checking completion (string from CLI) */
	minIterations: string;
	/** Maximum iterations, 0 for unlimited (string from CLI) */
	maxIterations: string;
	/** Phrase that signals loop completion */
	completionPromise: string;
	/** Optional agent name override */
	agent?: string;
}

/**
 * Starts a Ralph Wiggum iterative loop.
 * Parses and validates CLI options with Zod, then delegates to runLoop().
 * @param prompt - The task prompt for the loop
 * @param opts - Command options from CLI
 * @returns Resolves when the loop completes or is interrupted
 * @throws Exits process with code 1 if validation fails
 */
export async function loopCommand(
	prompt: string,
	opts: LoopOptions,
): Promise<void> {
	// Parse and validate options with Zod
	const result = LoopConfigSchema.safeParse({
		prompt,
		minIterations: Number.parseInt(opts.minIterations, 10),
		maxIterations: Number.parseInt(opts.maxIterations, 10),
		completionPromise: opts.completionPromise,
		agentName: opts.agent ?? null,
	});

	if (!result.success) {
		// Format Zod error messages
		const errorMessages = result.error.issues
			.map((issue) => `  - ${issue.path.join(".")}: ${issue.message}`)
			.join("\n");
		log.error(pc.red(`Validation error:\n${errorMessages}`));
		process.exit(1);
	}

	await runLoop(result.data);
}
